### ISSUE TITLE
C99 "snprintf" instead of _snprintf all over the place for GCC, Clang and VCR 2015

### ISSUE THREAD

**Issue by jasonwilliams200OK** [dotnet/coreclr#1058]
VC 2015 has (finally) brought support for standard C99 `snprintf`. There are several instances of `_snprintf` in this repo.

There is one difference between non-std `_snprintf` and std `snprintf` that the former does not add null-terminator at the end.

Instead of choosing `_snprintf` as a fallback, the correct solution is to have something what has been done in libuv repo:

[https://github.com/libuv/libuv/blob/../test/task.h#L187-L209](https://github.com/libuv/libuv/blob/9e7568ea7b2b9375be6563c9fb5e795c9753287e/test/task.h#L187-L209) union https://github.com/libuv/libuv/pull/369/files:

```c
/* Define inline for MSVC<2015 */ 
# if defined(_MSC_VER) && _MSC_VER < 1900 
#  define inline __inline 
# endif 
 
# if defined(_MSC_VER) && _MSC_VER < 1900 
/* Emulate snprintf() on MSVC<2015, _snprintf() doesn't zero-terminate the buffer 
 * on overflow... 
 */ 
inline int snprintf(char* buf, size_t len, const char* fmt, ...) { 
va_list ap; 
int n; 

 
va_start(ap, fmt); 
n = _vsprintf_p(buf, len, fmt, ap); 
va_end(ap); 

/* It's a sad fact of life that no one ever checks the return value of 
 * snprintf(). Zero-terminating the buffer hopefully reduces the risk 
 * of gaping security holes. 
 */ 
if (n < 0) 
  if (len > 0) 
    buf[0] = '\0'; 

 
  return n; 
} 
# endif 
```

We can probably define a proper equivalent (with null-terminator) fallback for VC < 2015 once and then use it everywhere else.

Inviting @janvorli

**Comment by janvorli** [comment]
@jasonwilliams200OK I don't see us using _snprintf in the coreclr, only the _snprintf_s, which always appends the zero terminator.
The only place (not counting PAL tests) where the _snprintf is being used is in the PAL in _safecrt_cfltcvt where the result of the function is checked and an empty string is returned if the result didn't fit the buffer.

**Comment by jasonwilliams200OK** [comment]
@janvorli, thanks for `_snprintf_s` tip.
I was actually using GitHub search and it seems like there are quite many instances where snprintf and its fallbacks are used. At one place even this kind of code is used:

```c
#ifdef PLATFORM_UNIX
# undef snprintf
```

Will it make sense to declare it in common utils and use one name `snprintf` everywhere else? Something like:

```c
# if defined(_MSC_VER) && _MSC_VER < 1900
# define snprintf _snprintf_s
# endif
```

**Comment by janvorli** [comment]
@jasonwilliams200OK The code that undefs the snprintf is obsolete, it happens at the end of  the src\pal\src\misc\perftrace.cpp file and so it has no effect. In fact, that file is not even being compiled.
I have full text searched the codebase again and I can only see the _snprintf being used in the PAL tests, where it doesn't really matter and in that single function in PAL where the return value is properly checked. 
Please note that PAL is used only on non-Windows systems and that we are not using the platform _snprintf. We have our own implementation. @jkotas, do you know why we don't call the platform _snprintf from the PAL _snprintf instead? Is it because we want to support VC++ specific format specifiers? 

@jkotas, do you see the perftrace.cpp stuff as something we will want to revive eventually?

**Comment by jkotas** [comment]
Yes, all printf-style formatting is overridden in the PAL. The primary reason is to handle Unicode as UTF16, but a few more quirks were added on top of it.

perftrace.cpp:  It is a poor man's profiler for PAL that is trying to log how much time is spent in individual PAL function. I doubt that it will be useful because of it is scoped to PAL only. We care about the performance of the whole system, not just PAL. There are better profiling/tracing tools these days for that.

**Comment by adityamandaleeka** [comment]
Reopening... we discussed this further in #1959 and it's worth looking into using the standard snprintf and removing support for VS2013 and older.
